### PR TITLE
feat: force user agent

### PR DESCRIPTION
**Pull request by Benoit382**
### What
Force user agent and update some dependencies

### Issues
#48

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ jobs:
       max-parallel: 1
       fail-fast: false
       matrix:
-        php: ['7.2', '7.3', '7.4', '8.0']
+        php: ['8.1', '8.2']
```
Comment (Dwarfex) [inline .github/workflows/ci.yml]: I would suggest not activly dropping support of Php 8.0 as it is still maintained and i would see this as a long living library.

File: src/Api.php [hunk 2]
```diff
@@ -129,12 +116,13 @@ class Api
...
      *
      * @param string $api the environment to search
-     * @param string $geography this parameter represent the the country  code and the interface of the language
+     * @param string $geography this parameter represent the the country code and the interface of the language
      * @param LoggerInterface $logger this parameter define an logger
      * @param ClientInterface|null $clientInterface
      * @param CacheInterface|null $cacheInterface
      */
     public function __construct(
+        public readonly string $userAgent,
```
Comment (Dwarfex) [inline src/Api.php]: you are breaking semantic versioning by adding this parameter in the front. 

We should aswell set a valuable default here, if we intend to use it.

File: src/Api.php [hunk 3]
```diff
@@ -368,7 +355,14 @@ public function downloadData(string $filePath, string $fileType = "mongodb")
...
         $url        = $this->buildUrl('data', self::FILE_TYPE_MAP[$fileType]);
         try {
-            $response = $this->httpClient->request('get', $url, ['sink' => $filePath]);
+            $response = $this->httpClient->request(
+                'get',
+                $url,
+                [
+                    'sink' => $filePath,
+                    'headers' => $this->getDefaultHeaders()
+                ]
```
Comment (Dwarfex) [inline src/Api.php]: I'd suggest setting up the clients default headers within the constructor - and not for each individual call

File: src/Api.php [hunk 4]
```diff
@@ -428,6 +420,7 @@ private function fetch(string $url, bool $isJsonFile = true): array
         }
         $this->logger->info('OpenFoodFact - fetch - GET : ' . $url . ' - ' . $response->getStatusCode());
 
+        /** @var array $jsonResult */
```
Comment (Dwarfex) [inline src/Api.php]: json_decode returns "mixed" not "array" - this is a wrong typehint

File: src/Api.php [hunk 5]
```diff
@@ -479,6 +473,7 @@ private function fetchPost(string $url, array $postData, bool $isMultipart = fal
 
         $this->logger->info('OpenFoodFact - fetch - GET : ' . $url . ' - ' . $response->getStatusCode());
 
+        /** @var array $jsonResult */
```
Comment (Dwarfex) [inline src/Api.php]: here aswell

File: src/Api.php [hunk 6]
```diff
@@ -492,34 +487,43 @@ private function fetchPost(string $url, array $postData, bool $isMultipart = fal
...
+                    throw new BadRequestException('Impossible to build requested url');
+                }
                 $baseUrl = implode('/', [
                   $this->geoUrl,
                   $service,
                   'v0',
-                  $resourceType,
-                  $parameters
+                  $resourceType ?? '',
+                  $parameters ?? ''
```
Comment (Dwarfex) [inline src/Api.php]: Isn't this null check useless as you verify in L498-500 for a given string?

File: src/Api.php [hunk 7]
```diff
@@ -492,34 +487,43 @@ private function fetchPost(string $url, array $postData, bool $isMultipart = fal
...
                 break;
             case 'data':
+                if(!is_string($resourceType ?? '')) {
+                    throw new BadRequestException('Impossible to build requested url');
+                }
                 $baseUrl = implode('/', [
                   $this->geoUrl,
                   $service,
-                  $resourceType
+                  $resourceType ?? ''
```
Comment (Dwarfex) [inline src/Api.php]:
```suggestion
                  $resourceType
```

File: src/Api.php [hunk 8]
```diff
@@ -492,34 +487,43 @@ private function fetchPost(string $url, array $postData, bool $isMultipart = fal
...
                 break;
             case 'cgi':
+                if(!is_string($resourceType ?? '')) {
+                    throw new BadRequestException('Impossible to build requested url');
+                }
                 $baseUrl = implode('/', [
                   $this->geoUrl,
                   $service,
-                  $resourceType
+                  $resourceType ?? ''
```
Comment (Dwarfex) [inline src/Api.php]:
```suggestion
                  $resourceType
```

File: src/Api.php [hunk 9]
```diff
@@ -531,17 +535,43 @@ private function buildUrl(string $service = null, $resourceType = null, $paramet
...
                     //need test
                     $resourceType = implode('/', ["state",  "ingredients-completed"]);
-                    $parameters   = 1;
+                    $parameters   = '1';
                 }
                 $baseUrl = implode('/', array_filter([
                     $this->geoUrl,
                     $resourceType,
-                    $parameters
+                    is_array($parameters) ? '' : $parameters
```
Comment (Dwarfex) [inline src/Api.php]: What is your intention here ? - Removing the parameters if they are an array? - Why?

File: src/Api.php [hunk 10]
```diff
@@ -531,17 +535,43 @@ private function buildUrl(string $service = null, $resourceType = null, $paramet
...
                     return !empty($value);
                 }));
                 break;
         }
         return $baseUrl;
     }
+
+    /**
+     * @return array
+     */
```
Comment (Dwarfex) [inline src/Api.php]:
```suggestion
```

File: src/Api.php [hunk 11]
```diff
@@ -531,17 +535,43 @@ private function buildUrl(string $service = null, $resourceType = null, $paramet
...
+    private function getDefaultOptions(): array
+    {
+        $data = [
+            'headers' => $this->getDefaultHeaders(),
```
Comment (Dwarfex) [inline src/Api.php]:
```suggestion
            'headers' => ['User-Agent' => 'SDK PHP - ' . $this->userAgent],
```

File: src/Api.php [hunk 12]
```diff
@@ -531,17 +535,43 @@ private function buildUrl(string $service = null, $resourceType = null, $paramet
...
+    /**
+     * @return array
+     */
+    private function getDefaultHeaders(): array
+    {
+        return [
+            'User-Agent' => 'SDK PHP - ' . $this->userAgent,
+        ];
+    }
+
```
Comment (Dwarfex) [inline src/Api.php]:
```suggestion
```

File: src/Collection.php [hunk 13]
```diff
@@ -103,38 +103,35 @@ public function searchCount(): int
...
     }
     /**
      * @inheritDoc
-     * @return Document|false
      */
-    public function next()
+    public function next(): void
     {
-        return next($this->listDocuments);
+        next($this->listDocuments);
```
Comment (Dwarfex) [inline src/Collection.php]: This looks right - but changes the interface for anyone using it.

File: src/Document.php [hunk 14]
```diff
@@ -18,22 +18,15 @@ class Document
...
-     */
-    private $api;
 
     /**
      * Initialization the document and specify from which API it was extract
      * @param array $data the whole data
-     * @param string|null $api the api name
      */
-    public function __construct(array $data, string $api = null)
+    public function __construct(array $data)
```
Comment (Dwarfex) [inline src/Document.php]: This is version breaking

File: tests/ApiFoodCacheTest.php [hunk 15]
```diff
@@ -28,12 +25,12 @@ protected function setUp(): void
             CURLOPT_FRESH_CONNECT => true,
             'defaults' => [
                 'headers' => [
-                    'CURLOPT_USERAGENT' => 'OFF - PHP - SDK - Unit Test',
+                    CURLOPT_USERAGENT => 'OFF - PHP - SDK - Unit Test',
                 ],
             ],
         ]);
```
Comment (Dwarfex) [inline tests/ApiFoodCacheTest.php]: Why not using the newly added user-agent here?

File: tests/FilesystemTrait.php [hunk 16]
```diff
@@ -1,17 +1,17 @@
 <?php
 
-namespace OpenFoodFacts;
+namespace OpenFoodFactsTests;
 
 trait FilesystemTrait
 {
-    public function recursiveDeleteDirectory(string $dir): void
+    public static function recursiveDeleteDirectory(string $dir): void
```
Comment (Dwarfex) [inline tests/FilesystemTrait.php]: again version breaking - and as this is a trait -> tests subfolder is the wrong location

File: tests/FilesystemTrait.php [hunk 17]
```diff
@@ -20,4 +20,10 @@ public function recursiveDeleteDirectory(string $dir): void
             rmdir($dir);
         }
     }
+    public static function cleanTestFolder(): void
+    {
+        self::recursiveDeleteDirectory('tests/tmp');
+        mkdir('tests/tmp', 0755);
+        mkdir('tests/tmp/cache', 0755);
+    }
```
Comment (Dwarfex) [inline tests/FilesystemTrait.php]: Why is this test related function in a generic trait?

**Review by Dwarfex — CHANGES_REQUESTED** [review]
This PR does a lot of things - besides adding a User Agent - and it breaks semantic versionig -> would result in the need of a new major version vor the library.

**Comment by teolemon** [comment]
@Benoit382 merge conflicts

**Comment by Benoit382** [comment]
Oups, i removed it :stuck_out_tongue: 
I will rework on it soon

**Pull request closed without merging**